### PR TITLE
Search by tags

### PR DESCRIPTION
**Pull request by Aeropio**

### CODE REVIEW

File: app/controllers/tag_listing_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class TagListingController < ApplicationController
+
+  def tag_list
```
Comment (chandra-seanergy) [inline app/controllers/tag_listing_controller.rb:3]: No need of extra API search case should be managed in Tag List API please remove this method

File: spec/controllers/tag_listing_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (chandra-seanergy) [inline spec/controllers/tag_listing_controller_spec.rb:1]: Please don't push the unnecessary files

File: app/assets/stylesheets/tag_listing.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the tag_listing controller here.
```
Comment (chandra-seanergy) [inline app/assets/stylesheets/tag_listing.scss:1]: Please don't push the unnecessary files

File: app/controllers/documents_controller.rb [hunk 4]
```diff
@@ -7,13 +7,25 @@ class DocumentsController < ApplicationController
   # Return document data or header filter data on basis of search text and filters selected along with total documents count matching criteria 
   def index
   	if params[:group].empty?
-  		if params[:search_text] and !params[:search_text].empty?
-  			documents, total_count, doc_arr=Document.search_text(params[:search_text], current_page(params[:skip]), current_user, params[:sort], params[:requireTotalCount], @apply, @exclude)
+      #only search text
```
Comment (chandra-seanergy) [inline app/controllers/documents_controller.rb:10]: can we manage the cases by switch cases instead of multiple if else statements

File: config/routes.rb [hunk 5]
```diff
@@ -82,5 +82,10 @@
     end
   end
   resources :tags
+  resources :tag_listing do
```
Comment (chandra-seanergy) [inline config/routes.rb:85]: please remove the routes no need for that

File: app/helpers/tag_listing_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module TagListingHelper
```
Comment (chandra-seanergy) [inline app/helpers/tag_listing_helper.rb:1]: Please remove the unnecessary files

File: app/models/tag.rb [hunk 7]
```diff
@@ -1,3 +1,8 @@
 class Tag < ApplicationRecord
   belongs_to :user
+
+  def self.find_tags(search_params)
```
Comment (chandra-seanergy) [inline app/models/tag.rb:4]: Please remove the method

File: app/controllers/tag_listing_controller.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    tags = Tag.find_tags(tag_list_params)
+    render json: {status: 200, message: "Tag list feched successfully.", tags: tags}
+  end
+
+  private
```
Comment (chandra-seanergy) [inline app/controllers/tag_listing_controller.rb:8]: Please make separate PR for different functionalities

File: app/controllers/documents_controller.rb (hunk 4 above)
Comment (Aeropio) [reply app/controllers/documents_controller.rb:10]: switch statements can be used if we are using same variable across , but here in this case we cannot use

File: app/controllers/documents_controller.rb [hunk 9]
```diff
@@ -7,13 +7,25 @@ class DocumentsController < ApplicationController
...
+        documents, total_count, doc_arr=Document.search_text(params[:search_text], current_page(params[:skip]),
+          current_user, params[:sort], params[:requireTotalCount], @apply, @exclude, params[:tag_ids])
   		else
-  			documents, total_count, doc_arr=Document.document_details(current_page(params[:skip]), current_user, params[:sort], params[:requireTotalCount], @apply, @exclude)
+  			documents, total_count, doc_arr=Document.document_details(current_page(params[:skip]), current_user,
+          params[:sort], params[:requireTotalCount], @apply, @exclude)
   		end
   	else
-  		documents, total_count, doc_arr=Document.get_context_filters(current_page(params[:skip]), params[:group] , params[:requireTotalCount], params[:filter], params[:applied_filter], params[:exclude_filter], params[:search_text])
+  		documents, total_count, doc_arr=Document.get_context_filters(current_page(params[:skip]), params[:group] ,
```
Comment (DhruvRSeanergy) [inline app/controllers/documents_controller.rb:27]: need to include tag ids while fetching filter contents too as filters are dependent on previous actions

File: app/controllers/documents_controller.rb [hunk 10]
```diff
@@ -7,13 +7,25 @@ class DocumentsController < ApplicationController
...
   def index
   	if params[:group].empty?
-  		if params[:search_text] and !params[:search_text].empty?
-  			documents, total_count, doc_arr=Document.search_text(params[:search_text], current_page(params[:skip]), current_user, params[:sort], params[:requireTotalCount], @apply, @exclude)
+      #only search text
+  		if params[:search_text] and !params[:search_text].empty? and params[:tag_ids].empty?
+  			documents, total_count, doc_arr=Document.search_text(params[:search_text], current_page(params[:skip]),
+          current_user, params[:sort], params[:requireTotalCount], @apply, @exclude)
+      #only tags
+      elsif params[:search_text].empty? and params[:tag_ids].present?
```
Comment (DhruvRSeanergy) [inline app/controllers/documents_controller.rb:15]: Try to manage tags condition in model so as not to make the controller methods bulky and also avoid so many if else checks and try to handle the case using ternaries and in ActiveRecords query

File: app/models/document.rb [hunk 11]
```diff
@@ -7,20 +7,28 @@ class Document < ApplicationRecord
   has_many :user_selections, foreign_key: :doc_id
 
 	# Searches all documents on the basis of filters applied and sorts it on the basis of sort column and order selected
-  def self.document_details(page, user, sort_on, requireTotalCount, apply, exclude)
+  def self.document_details(page, user, sort_on, requireTotalCount, apply, exclude, tag_params="")
   	where_clause=(self.apply_filter(apply, exclude).split('and')[1..-1].join('and') rescue"")
+    if tag_params.present?
+      if where_clause.present?
+        where_clause = where_clause+" and documents.id in (select document_id from tag_documents where tag_id in (#{tag_params}))"
```
Comment (DhruvRSeanergy) [inline app/models/document.rb:14]: try if you can reduce the number of lines of codes

File: app/models/document.rb [hunk 12]
```diff
@@ -80,7 +88,8 @@ def self.preview_docs_attr(document, doc_ids, pages)
   end
 
   # returns an array of hashes where each hash contains document's attributes that would be consumed by the front end to populate content on the front end
-	def self.document_attrs(documents, user, total_count, pages=[])
+	def self.document_attrs(documents, user, total_count, tag_params="", pages=[])
+    tag_ids = tag_params.split(",").map { |s| s.to_i }
```
Comment (DhruvRSeanergy) [inline app/models/document.rb:92]: follow the indentation
